<h5>Importing libraries</h5>

In [ ]:
import pandas as pd
import numpy as np
from scipy.stats import f_oneway
import warnings
warnings.filterwarnings('ignore')

<h3>Importing dataset and converting it into column</h3>

In [ ]:
uk = pd.read_csv('/content/drive/MyDrive/Colab Notebooks/datasets/uk_road_accident.csv')

In [ ]:
uk

,Index,Accident_Severity,Accident Date,Latitude,Light_Conditions,District Area,Longitude,Number_of_Casualties,Number_of_Vehicles,Road_Surface_Conditions,Road_Type,Urban_or_Rural_Area,Weather_Conditions,Vehicle_Type
0,200701BS64157,Serious,5/6/2019,51.506187,Darkness - lights lit,Kensington and Chelsea,-0.209082,1,2,Dry,Single carriageway,Urban,Fine no high winds,Car
1,200701BS65737,Serious,2/7/2019,51.495029,Daylight,Kensington and Chelsea,-0.173647,1,2,Wet or damp,Single carriageway,Urban,Raining no high winds,Car
2,200701BS66127,Serious,26-08-2019,51.517715,Darkness - lighting unknown,Kensington and Chelsea,-0.210215,1,3,Dry,NaN,Urban,NaN,Taxi/Private hire car
3,200701BS66128,Serious,16-08-2019,51.495478,Daylight,Kensington and Chelsea,-0.202731,1,4,Dry,Single carriageway,Urban,Fine no high winds,Bus or coach (17 or more pass seats)
4,200701BS66837,Slight,3/9/2019,51.488576,Darkness - lights lit,Kensington and Chelsea,-0.192487,1,2,Dry,NaN,Urban,NaN,Other vehicle
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
660674,201091NM01760,Slight,18-02-2022,57.374005,Daylight,Highland,-3.467828,2,1,Dry,Single carriageway,Rural,Fine no high winds,Car
660675,201091NM01881,Slight,21-02-2022,57.232273,Darkness - no lighting,Highland,-3.809281,1,1,Frost or ice,Single carriageway,Rural,Fine no high winds,Car
660676,201091NM01935,Slight,23-02-2022,57.585044,Daylight,Highland,-3.862727,1,3,Frost or ice,Single carriageway,Rural,Fine no high winds,Car
660677,201091NM01964,Serious,23-02-2022,57.214898,Darkness - no lighting,Highland,-3.823997,1,2,Wet or damp,Single carriageway,Rural,Fine no high winds,Motorcycle over 500cc


In [ ]:
uk.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 660679 entries, 0 to 660678
Data columns (total 14 columns):
 #   Column                   Non-Null Count   Dtype  
---  ------                   --------------   -----  
 0   Index                    660679 non-null  object 
 1   Accident_Severity        660679 non-null  object 
 2   Accident Date            660679 non-null  object 
 3   Latitude                 660654 non-null  float64
 4   Light_Conditions         660679 non-null  object 
 5   District Area            660679 non-null  object 
 6   Longitude                660653 non-null  float64
 7   Number_of_Casualties     660679 non-null  int64  
 8   Number_of_Vehicles       660679 non-null  int64  
 9   Road_Surface_Conditions  659953 non-null  object 
 10  Road_Type                656159 non-null  object 
 11  Urban_or_Rural_Area      660664 non-null  object 
 12  Weather_Conditions       646551 non-null  object 
 13  Vehicle_Type             660679 non-null  object 
dtypes: f

<h3>Changing the data types of the columns</h3>

In [ ]:
uk.dtypes

,0
Index,object
Accident_Severity,object
Accident Date,object
Latitude,float64
Light_Conditions,object
District Area,object
Longitude,float64
Number_of_Casualties,int64
Number_of_Vehicles,int64
Road_Surface_Conditions,object


In [ ]:
uk['Index'] = uk['Index'].astype('category')
uk['Accident_Severity'] = uk['Accident_Severity'].astype('category')
uk['Accident Date'] = uk['Accident Date'].astype('category')
uk['Light_Conditions'] = uk['Light_Conditions'].astype('category')
uk['District Area'] = uk['District Area'].astype('category')
uk['Road_Surface_Conditions'] = uk['Road_Surface_Conditions'].astype('category')
uk['Road_Type'] = uk['Road_Type'].astype('category')
uk['Urban_or_Rural_Area'] = uk['Urban_or_Rural_Area'].astype('category')
uk['Weather_Conditions'] = uk['Weather_Conditions'].astype('category')
uk['Vehicle_Type'] = uk['Vehicle_Type'].astype('category')

In [ ]:
uk.dtypes

,0
Index,category
Accident_Severity,category
Accident Date,category
Latitude,float64
Light_Conditions,category
District Area,category
Longitude,float64
Number_of_Casualties,int64
Number_of_Vehicles,int64
Road_Surface_Conditions,category


<h3>Cleaning empty datas</h3>

In [ ]:
uk.isna().sum()

,0
Index,0
Accident_Severity,0
Accident Date,0
Latitude,25
Light_Conditions,0
District Area,0
Longitude,26
Number_of_Casualties,0
Number_of_Vehicles,0
Road_Surface_Conditions,726


<h5>Cleaning empty numerical datas</h5>

In [ ]:
uk['Latitude'] = uk['Latitude'].fillna(uk['Latitude'].mean())
uk['Longitude'] = uk['Longitude'].fillna(uk['Longitude'].mean())

In [ ]:
uk.isna().sum()

,0
Index,0
Accident_Severity,0
Accident Date,0
Latitude,0
Light_Conditions,0
District Area,0
Longitude,0
Number_of_Casualties,0
Number_of_Vehicles,0
Road_Surface_Conditions,726


<h5>Cleaning non-numerical empty datas </h5>

In [ ]:
uk['Road_Surface_Conditions'] = uk['Road_Surface_Conditions'].fillna(uk['Road_Surface_Conditions'].mode()[0])
uk['Road_Type'] = uk['Road_Type'].fillna(uk['Road_Type'].mode()[0])
uk['Urban_or_Rural_Area'] = uk['Urban_or_Rural_Area'].fillna(uk['Urban_or_Rural_Area'].mode()[0])
uk['Weather_Conditions'] = uk['Weather_Conditions'].fillna(uk['Weather_Conditions'].mode()[0])

In [ ]:
uk.isna().sum()

,0
Index,0
Accident_Severity,0
Accident Date,0
Latitude,0
Light_Conditions,0
District Area,0
Longitude,0
Number_of_Casualties,0
Number_of_Vehicles,0
Road_Surface_Conditions,0


<h1>Exploratory Data Analytics</h5>

<h1>1. What percentages do each accident severity represent on the accident report? and which severity had the highest percentage on the report.

In [ ]:
uk['Accident_Severity'].unique()

['Serious', 'Slight', 'Fatal']
Categories (3, object): ['Fatal', 'Serious', 'Slight']

In [ ]:
severity = uk['Accident_Severity']

In [ ]:
Slight, Serious, Fatal = severity.value_counts()

<p>Percentage of Slight category</p>

In [ ]:
(Slight / len(uk)) * 100

85.33660067899842

<p>Slight severity consumes <strong>85%</strong> of overall reports</p>

<p>Percentage of serious category</p>

In [ ]:
(Serious / len(uk)) * 100

13.352475256516403

<p>Serious severity consumes <strong>13%</strong> of overall reports</p>

In [ ]:
(Fatal / len(uk)) * 100

1.3109240644851736

<p>Fatal severity consumes <strong>1%</strong> of overall reports</p>

<h1>Insight # 1: With the calculation above, we found out that Serious accident severity consumes 85.33% of the overall reportts while Serious and Fatal have 13.35% and 1.31% respectively</h1>

<h1>2. Does the weather condition and vehicle type affect the number of casualties? If so, which weather condition and vehicle type recorded the highest number of casualties?

In [ ]:
 pd.set_option('display.max_colwidth', 500)

In [ ]:
weatherandvehicle = uk.groupby(['Weather_Conditions', "Vehicle_Type"])['Number_of_Casualties'].size()

In [ ]:
uk['Weather_Conditions'].value_counts()

,count
Weather_Conditions,
Fine no high winds,535013
Raining no high winds,79696
Other,17150
Raining + high winds,9615
Fine + high winds,8554
Snowing no high winds,6238
Fog or mist,3528
Snowing + high winds,885


<h3>Breakdown</h3>

In [ ]:
# Fine and no high winds
weatherandvehicle['Fine no high winds'].sort_values(ascending = False)

,Number_of_Casualties
Vehicle_Type,
Car,403324
Van / Goods 3.5 tonnes mgw or under,27603
Bus or coach (17 or more pass seats),20963
Motorcycle over 500cc,20793
Goods 7.5 tonnes mgw and over,13938
Motorcycle 125cc and under,12407
Taxi/Private hire car,10733
Motorcycle over 125cc and up to 500cc,6228
Motorcycle 50cc and under,6165


In [ ]:
# raining no high winds
weatherandvehicle['Raining no high winds'].sort_values(ascending = False)

,Number_of_Casualties
Vehicle_Type,
Car,59940
Van / Goods 3.5 tonnes mgw or under,4203
Bus or coach (17 or more pass seats),3182
Motorcycle over 500cc,3135
Goods 7.5 tonnes mgw and over,2114
Motorcycle 125cc and under,1830
Taxi/Private hire car,1622
Motorcycle 50cc and under,897
Motorcycle over 125cc and up to 500cc,893


In [ ]:
# raining+ high winds
weatherandvehicle['Raining + high winds'].sort_values(ascending = False)

,Number_of_Casualties
Vehicle_Type,
Car,7206
Van / Goods 3.5 tonnes mgw or under,508
Bus or coach (17 or more pass seats),389
Motorcycle over 500cc,354
Goods 7.5 tonnes mgw and over,275
Motorcycle 125cc and under,220
Taxi/Private hire car,190
Motorcycle over 125cc and up to 500cc,117
Motorcycle 50cc and under,107


In [ ]:
# Fine + high winds
weatherandvehicle['Fine + high winds'].sort_values(ascending = False)

,Number_of_Casualties
Vehicle_Type,
Car,6463
Van / Goods 3.5 tonnes mgw or under,408
Bus or coach (17 or more pass seats),350
Motorcycle over 500cc,315
Goods 7.5 tonnes mgw and over,251
Taxi/Private hire car,187
Motorcycle 125cc and under,180
Motorcycle 50cc and under,106
Goods over 3.5t. and under 7.5t,94


In [ ]:
# snowing no high winds
weatherandvehicle['Snowing no high winds'].sort_values(ascending = False)

,Number_of_Casualties
Vehicle_Type,
Car,4748
Van / Goods 3.5 tonnes mgw or under,296
Motorcycle over 500cc,254
Bus or coach (17 or more pass seats),219
Goods 7.5 tonnes mgw and over,149
Motorcycle 125cc and under,146
Taxi/Private hire car,136
Motorcycle 50cc and under,75
Motorcycle over 125cc and up to 500cc,72


In [ ]:
# Fog or mist
weatherandvehicle['Fog or mist'].sort_values(ascending = False)

,Number_of_Casualties
Vehicle_Type,
Car,2641
Van / Goods 3.5 tonnes mgw or under,192
Bus or coach (17 or more pass seats),134
Motorcycle over 500cc,118
Goods 7.5 tonnes mgw and over,93
Motorcycle 125cc and under,82
Taxi/Private hire car,78
Motorcycle 50cc and under,50
Motorcycle over 125cc and up to 500cc,45


In [ ]:
# Snowing + high winds
weatherandvehicle['Snowing + high winds'].sort_values(ascending = False)

,Number_of_Casualties
Vehicle_Type,
Car,677
Van / Goods 3.5 tonnes mgw or under,48
Motorcycle over 500cc,33
Bus or coach (17 or more pass seats),32
Goods 7.5 tonnes mgw and over,28
Taxi/Private hire car,20
Motorcycle 125cc and under,16
Motorcycle over 125cc and up to 500cc,10
Motorcycle 50cc and under,6


<h1>Insight #2:</h1>
<h5>Car consistently dominated the vehucle type per casualties. It remained as the number one cause of accident regardlessof the weather condition. Surprisingly, contrary to the popular belief that wet roads causes more accidents, the data shows that 500000 of the overall report occured on a dry weather.

<h1>Question #3: Which is the safest? Urban or Rural area. Determine if the Light condition and Road type of an area determine the number of casualties it  will record. </h1>

In [ ]:
uk.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 660679 entries, 0 to 660678
Data columns (total 14 columns):
 #   Column                   Non-Null Count   Dtype   
---  ------                   --------------   -----   
 0   Index                    660679 non-null  category
 1   Accident_Severity        660679 non-null  category
 2   Accident Date            660679 non-null  category
 3   Latitude                 660679 non-null  float64 
 4   Light_Conditions         660679 non-null  category
 5   District Area            660679 non-null  category
 6   Longitude                660679 non-null  float64 
 7   Number_of_Casualties     660679 non-null  int64   
 8   Number_of_Vehicles       660679 non-null  int64   
 9   Road_Surface_Conditions  660679 non-null  category
 10  Road_Type                660679 non-null  category
 11  Urban_or_Rural_Area      660679 non-null  category
 12  Weather_Conditions       660679 non-null  category
 13  Vehicle_Type             660679 non-null  ca

In [ ]:
urbanorrural = uk.groupby(['Urban_or_Rural_Area','Light_Conditions','Road_Type'])['Number_of_Casualties'].count()

<h3>Breakdown</h3>

In [ ]:
# Rural
urbanorrural['Rural']

Light_Conditions             Road_Type         
Darkness - lighting unknown  Dual carriageway         512
                             One way street            15
                             Roundabout               188
                             Single carriageway      1702
                             Slip road                 50
Darkness - lights lit        Dual carriageway        6556
                             One way street           182
                             Roundabout              3026
                             Single carriageway     14248
                             Slip road                683
Darkness - lights unlit      Dual carriageway         254
                             One way street             6
                             Roundabout                71
                             Single carriageway       611
                             Slip road                 19
Darkness - no lighting       Dual carriageway        7019
                             One way street            74
                             Roundabout               172
                             Single carriageway     27926
                             Slip road                326
Daylight                     Dual carriageway       34374
                             One way street           916
                             Roundabout             12088
                             Single carriageway    124756
                             Slip road               3216
Name: Number_of_Casualties, dtype: int64

In [ ]:
urbanorrural['Urban']

Light_Conditions             Road_Type         
Darkness - lighting unknown  Dual carriageway         393
                             One way street           109
                             Roundabout               282
                             Single carriageway      3197
                             Slip road                 36
Darkness - lights lit        Dual carriageway       13686
                             One way street          3128
                             Roundabout              7501
                             Single carriageway     79691
                             Slip road                632
Darkness - lights unlit      Dual carriageway         211
                             One way street            52
                             Roundabout               118
                             Single carriageway      1186
                             Slip road                 15
Darkness - no lighting       Dual carriageway         441
                             One way street            29
                             Roundabout                52
                             Single carriageway      1367
                             Slip road                 31
Daylight                     Dual carriageway       35977
                             One way street          9048
                             Roundabout             20493
                             Single carriageway    241970
                             Slip road               2033
Name: Number_of_Casualties, dtype: int64

In [157]:
# Checking if Rural recorded more number of casualties than Urban
(urbanorrural['Rural'] > urbanorrural['Urban']).value_counts()

,count
Number_of_Casualties,
False,14
True,11


<p>We could say here that rural is safer than Urban, but we still need to compare the number of casualties each have</p>

In [158]:
urbanorrural['Rural'].sum()

np.int64(238990)

In [159]:
urbanorrural['Urban'].sum()

np.int64(421678)

<p>Now, let's determine which Road Type and Light condition from each area recorded the highest number of casualties

In [162]:
urbanorrural['Rural'].sort_values(ascending=False).head(n=5)

Light_Conditions        Road_Type         
Daylight                Single carriageway    124756
                        Dual carriageway       34374
Darkness - no lighting  Single carriageway     27926
Darkness - lights lit   Single carriageway     14248
Daylight                Roundabout             12088
Name: Number_of_Casualties, dtype: int64

In [163]:
urbanorrural['Urban'].sort_values(ascending=False).head(n=5)

Light_Conditions       Road_Type         
Daylight               Single carriageway    241970
Darkness - lights lit  Single carriageway     79691
Daylight               Dual carriageway       35977
                       Roundabout             20493
Darkness - lights lit  Dual carriageway       13686
Name: Number_of_Casualties, dtype: int64

<h1>Insight #3:</h1>
<p>Daylight recorded the highest number of casualties, this would mean that drivers were more careful at night that they don't take extra car during daytime. Also, single carriageway road types recorded the highest number of casualties

<h1>Question #4: Traffic is one of the main source of accidents, determine which district areas recorded an accident involving more than 3 cars</h1>

In [164]:
uk['District Area']

,count
Number_of_Vehicles,
2,391995
1,200787
3,52806
4,11300
5,2464
6,758
7,296
8,147
9,57
